File: backend/src/main/java/com/expense/expensetracker/service/ExpenseService.java
package com.expense.expensetracker.service;

import com.expense.expensetracker.entity.Expense;
import com.expense.expensetracker.repository.ExpenseRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ExpenseService {

    private final ExpenseRepository expenseRepository;

    public ExpenseService(ExpenseRepository expenseRepository) {
        this.expenseRepository = expenseRepository;
    }

    public Expense createExpense(Expense expense) {
        return expenseRepository.save(expense);
    }

    public List<Expense> getAllExpenses() {
        return expenseRepository.findAll();
    }

    public Expense getExpenseById(Long id) {
        return expenseRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Expense not found with id: " + id));
    }

    public Expense updateExpense(Long id, Expense updatedExpense) {

        Expense expense = getExpenseById(id);

        expense.setTitle(updatedExpense.getTitle());
        expense.setAmount(updatedExpense.getAmount());
        expense.setCategory(updatedExpense.getCategory());
        expense.setCurrency(updatedExpense.getCurrency());
        expense.setDate(updatedExpense.getDate());
        expense.setDescription(updatedExpense.getDescription());

        return expenseRepository.save(expense);
    }

    public void deleteExpense(Long id) {

        Expense expense = getExpenseById(id);
        expenseRepository.delete(expense);
    }

    public List<Expense> searchExpenses(String keyword) {
        return expenseRepository.findByTitleContainingIgnoreCase(keyword);
    }

    public List<Expense> getExpensesByCategory(String category) {
        return expenseRepository.findByCategoryIgnoreCase(category);
    }
}

File: frontend/src/services/expenseService.js
import axios from "axios";

const API_URL = "http://localhost:8080/api/expenses";

export const getExpenses = () => axios.get(API_URL);

export const getExpenseById = (id) =>
  axios.get(`${API_URL}/${id}`);

export const createExpense = (expense) =>
  axios.post(API_URL, expense);

export const updateExpense = (id, expense) =>
  axios.put(`${API_URL}/${id}`, expense);

export const deleteExpense = (id) =>
  axios.delete(`${API_URL}/${id}`);

export const searchExpenses = (keyword) =>
  axios.get(`${API_URL}/search?keyword=${keyword}`);

export const getExpensesByCategory = (category) =>
  axios.get(`${API_URL}/category/${category}`);